# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/gl/SharedSurfaceEGL.h
# Commit: bbfe23c61add
# Full Hash: bbfe23c61add07377daf80696fd6996c0463f5b6
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-09 09:21:34
# Regressor Bug: 1632249
# File Overlap Count: 1
# Description:
#   Bug 1632249 - Support compositing out-of-process WebGL. r=handyman,lsalzman,nical,geckoview-reviewers,agi,imanol
#   
#   * Majorly simplity CanvasRenderer
#   * Replace GLScreenBuffer with trivial GLSwapChain
#   * Use descriptor structs so that future SharedSurface changes aren't so painful
# ==============================================================================

diff -r 8fcfac450ecc -r bbfe23c61add gfx/gl/SharedSurfaceEGL.h
--- a/gfx/gl/SharedSurfaceEGL.h	Mon Jun 08 20:13:01 2020 +0000
+++ b/gfx/gl/SharedSurfaceEGL.h	Mon Jun 08 20:34:15 2020 +0000
@@ -18,49 +18,29 @@
 namespace mozilla {
 namespace gl {
 
-class GLContext;
 class GLLibraryEGL;
 
-class SharedSurface_EGLImage : public SharedSurface {
- public:
-  static UniquePtr<SharedSurface_EGLImage> Create(GLContext* prodGL,
-                                                  const GLFormats& formats,
-                                                  const gfx::IntSize& size,
-                                                  bool hasAlpha,
-                                                  EGLContext context);
+// -
+// EGLImage
 
-  static SharedSurface_EGLImage* Cast(SharedSurface* surf) {
-    MOZ_ASSERT(surf->mType == SharedSurfaceType::EGLImageShare);
-
-    return (SharedSurface_EGLImage*)surf;
-  }
-
-  static bool HasExtensions(GLLibraryEGL* egl, GLContext* gl);
-
- protected:
+class SharedSurface_EGLImage final : public SharedSurface {
   mutable Mutex mMutex;
-  const GLFormats mFormats;
-  GLuint mProdTex;
+  EGLSync mSync = 0;
 
  public:
   const EGLImage mImage;
 
- protected:
-  EGLSync mSync;
+  static UniquePtr<SharedSurface_EGLImage> Create(const SharedSurfaceDesc&);
 
-  SharedSurface_EGLImage(GLContext* gl, const gfx::IntSize& size, bool hasAlpha,
-                         const GLFormats& formats, GLuint prodTex,
-                         EGLImage image);
+ protected:
+  SharedSurface_EGLImage(const SharedSurfaceDesc&,
+                         UniquePtr<MozFramebuffer>&& fb, EGLImage);
 
   void UpdateProdTexture(const MutexAutoLock& curAutoLock);
 
  public:
   virtual ~SharedSurface_EGLImage();
 
-  virtual layers::TextureFlags GetTextureFlags() const override {
-    return layers::TextureFlags::DEALLOCATE_CLIENT;
-  }
-
   virtual void LockProdImpl() override {}
   virtual void UnlockProdImpl() override {}
 
@@ -70,95 +50,57 @@
   virtual void ProducerReadAcquireImpl() override;
   virtual void ProducerReadReleaseImpl() override{};
 
-  virtual GLuint ProdTexture() override { return mProdTex; }
-
-  // Implementation-specific functions below:
-  // Returns texture and target
-  virtual bool ToSurfaceDescriptor(
-      layers::SurfaceDescriptor* const out_descriptor) override;
+  Maybe<layers::SurfaceDescriptor> ToSurfaceDescriptor() override;
 
   virtual bool ReadbackBySharedHandle(
       gfx::DataSourceSurface* out_surface) override;
 };
 
-class SurfaceFactory_EGLImage : public SurfaceFactory {
+class SurfaceFactory_EGLImage final : public SurfaceFactory {
  public:
-  // Fallible:
-  static UniquePtr<SurfaceFactory_EGLImage> Create(
-      GLContext* prodGL, const SurfaceCaps& caps,
-      const RefPtr<layers::LayersIPCChannel>& allocator,
-      const layers::TextureFlags& flags);
+  static UniquePtr<SurfaceFactory_EGLImage> Create(GLContext&);
 
- protected:
-  const EGLContext mContext;
-
-  SurfaceFactory_EGLImage(GLContext* prodGL, const SurfaceCaps& caps,
-                          const RefPtr<layers::LayersIPCChannel>& allocator,
-                          const layers::TextureFlags& flags, EGLContext context)
-      : SurfaceFactory(SharedSurfaceType::EGLImageShare, prodGL, caps,
-                       allocator, flags),
-        mContext(context) {}
+ private:
+  explicit SurfaceFactory_EGLImage(const PartialSharedSurfaceDesc& desc)
+      : SurfaceFactory(desc) {}
 
  public:
-  virtual UniquePtr<SharedSurface> CreateShared(
-      const gfx::IntSize& size) override {
-    bool hasAlpha = mReadCaps.alpha;
-    return SharedSurface_EGLImage::Create(mGL, mFormats, size, hasAlpha,
-                                          mContext);
+  virtual UniquePtr<SharedSurface> CreateSharedImpl(
+      const SharedSurfaceDesc& desc) override {
+    return SharedSurface_EGLImage::Create(desc);
   }
 };
 
+// -
+// SurfaceTexture
+
 #ifdef MOZ_WIDGET_ANDROID
 
-class SharedSurface_SurfaceTexture : public SharedSurface {
+class SharedSurface_SurfaceTexture final : public SharedSurface {
+  const java::GeckoSurface::GlobalRef mSurface;
+  const EGLSurface mEglSurface;
+  EGLSurface mOrigEglSurface = 0;
+
  public:
   static UniquePtr<SharedSurface_SurfaceTexture> Create(
-      GLContext* prodGL, const GLFormats& formats, const gfx::IntSize& size,
-      bool hasAlpha, java::GeckoSurface::Param surface);
-
-  static SharedSurface_SurfaceTexture* Cast(SharedSurface* surf) {
-    MOZ_ASSERT(surf->mType == SharedSurfaceType::AndroidSurfaceTexture);
-
-    return (SharedSurface_SurfaceTexture*)surf;
-  }
+      const SharedSurfaceDesc&);
 
   java::GeckoSurface::Param JavaSurface() { return mSurface; }
 
  protected:
-  java::GeckoSurface::GlobalRef mSurface;
-  EGLSurface mEglSurface;
-  EGLSurface mOrigEglSurface;
-
-  SharedSurface_SurfaceTexture(GLContext* gl, const gfx::IntSize& size,
-                               bool hasAlpha, const GLFormats& formats,
-                               java::GeckoSurface::Param surface,
-                               EGLSurface eglSurface);
+  SharedSurface_SurfaceTexture(const SharedSurfaceDesc&,
+                               java::GeckoSurface::Param surface, EGLSurface);
 
  public:
   virtual ~SharedSurface_SurfaceTexture();
 
-  virtual layers::TextureFlags GetTextureFlags() const override {
-    return layers::TextureFlags::DEALLOCATE_CLIENT;
-  }
-
   virtual void LockProdImpl() override;
   virtual void UnlockProdImpl() override;
 
   virtual void ProducerAcquireImpl() override {}
   virtual void ProducerReleaseImpl() override {}
 
-  virtual void ProducerReadAcquireImpl() override {}
-  virtual void ProducerReadReleaseImpl() override {}
-
-  // Implementation-specific functions below:
-  // Returns texture and target
-  virtual bool ToSurfaceDescriptor(
-      layers::SurfaceDescriptor* const out_descriptor) override;
-
-  virtual bool ReadbackBySharedHandle(
-      gfx::DataSourceSurface* out_surface) override {
-    return false;
-  }
+  Maybe<layers::SurfaceDescriptor> ToSurfaceDescriptor() override;
 
   virtual void Commit() override;
 
@@ -167,25 +109,14 @@
   virtual bool IsBufferAvailable() const override;
 };
 
-class SurfaceFactory_SurfaceTexture : public SurfaceFactory {
+class SurfaceFactory_SurfaceTexture final : public SurfaceFactory {
  public:
-  // Fallible:
-  static UniquePtr<SurfaceFactory_SurfaceTexture> Create(
-      GLContext* prodGL, const SurfaceCaps& caps,
-      const RefPtr<layers::LayersIPCChannel>& allocator,
-      const layers::TextureFlags& flags);
+  explicit SurfaceFactory_SurfaceTexture(GLContext&);
 
- protected:
-  SurfaceFactory_SurfaceTexture(
-      GLContext* prodGL, const SurfaceCaps& caps,
-      const RefPtr<layers::LayersIPCChannel>& allocator,
-      const layers::TextureFlags& flags)
-      : SurfaceFactory(SharedSurfaceType::AndroidSurfaceTexture, prodGL, caps,
-                       allocator, flags) {}
-
- public:
-  virtual UniquePtr<SharedSurface> CreateShared(
-      const gfx::IntSize& size) override;
+  virtual UniquePtr<SharedSurface> CreateSharedImpl(
+      const SharedSurfaceDesc& desc) override {
+    return SharedSurface_SurfaceTexture::Create(desc);
+  }
 };
 
 #endif  // MOZ_WIDGET_ANDROID